Update EIP-7668: Move to Draft (#9538)
ethereum/EIPs: EIPS/eip-7668.md

eip: 7668
title: Remove bloom filters
description: Remove bloom filters from the execution block
author: Vitalik Buterin (@vbuterin), Andrew Day (@acolytec3)
discussions-to: https://ethereum-magicians.org/t/eip-7653-remove-bloom-filters/19447

> status: Stagnant

Reviewer: ```suggestion
status: Draft
```

## Abstract

Require the bloom filters in an execution block, including at the top level and in the receipt object, to be hardcoded to `0xffff...ff`.

## Motivation

Logs were originally introduced to give applications a way to record information about onchain events, which decentralized applications (dapps) would be able to easily query. Using bloom filters, dapps would be able to quickly go through the history, identify the few blocks that contained logs relative to their application, and then quickly identify which individual transactions have the logs that they need.

> In practice, this mechanism is far too slow. Almost all dapps that access history end up doing so not through RPC calls to an Ethereum node (even a remote-hosted one), but through centralized extra-protocol services.  Not too mention, in a chain with relatively full blocks, the false positive rate in bloom filters has increased to the point of making bloom filters almost useless.

Reviewer 2: ```suggestion
In practice, this mechanism is far too slow. Almost all dapps that access history end up doing so not through RPC calls to an Ethereum node (even a remote-hosted one), but through centralized extra-protocol services.  Not to mention, in a chain with relatively full blocks, the false positive rate in bloom filters has increased to the point of making bloom filters almost useless.
```

One could also point to the fact that the gas limit will likely rise to 60M, which means more txs are packed in the block and which would thus bloat the filter even more because of more gas available for LOGs.

> This EIP proposes recognizing that reality, and removing bloom filters from the protocol since it places an unnecessary compute burden on execution clients computing what is for all prati. Applications that need history querying would be encouraged to develop and use decentralized protocols that create provable log indexes using eg. ZK-SNARKs and incrementally-verifiable computation.

Reviewer 2: This sentence seems incomplete?

Reviewer 3: Then complet it so it can balance

Author: > This sentence seems incomplete?

Yup, sorry, typo-d it.  Have updated with my latest commit

## Specification

> The `logsbloom` field of the execution block header and the `logsbloom` field of the transaction receipt is now required to be `0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` (ie. 32 bytes filled with `ff`).

Reviewer 2: ```suggestion
The `logsBloom` field of the execution block header and the `logsBloom` field of the transaction receipt is now required to be `0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff` (ie. 256 bytes filled with `ff`).
```

Reviewer 2: I guess cut it off (the 0xffff..ff), because this is wayyy too long.

Also, I think that the original (empty bytes it instead of writing the 1s) is better for storage and performance reasons. I know we wanted to write the all-bits-1 version because then the bloom filter would be positive for any query, but I think this can be handled at the application side (RPC). We could even wrap the RPC such that it returns the 0xffff..ffff as `logsBloom` for instance, or just even better really deprecate `logsBloom` and thus writing empty bytes.

We could also consider completely removing the `logsBloom` field from the receipts. This would save some extra bytes when hashing the receipt for the `receiptsTrie`.

Author: Are you saying replace the current field with a static byte array of 256 bytes of 0 or just an empty array?   I'm not confident on the implications for the RLP encoding of a tx receipt or block header.  I guess maybe nothing if we say as of the fork block, the logsBloom field becomes a `0` or `[]`, right?

Reviewer 2: If we input the empty bytes for the `logsBloom` starting at fork block then the RLP structure will not change (besides that the field is now not the constant 256 bytes long, but now 0. This is not a problem, we have variable-length fields like `extraData` which can be something between 0-32 bytes). If we would input it as empty list then this would (from my perspective) change the structure somewhat. I'd say empty bytes is the way to go. So empty bytes is not the empty list `[]`, it is a field of length 0. (This is possible, txs with a `to` field of 0 bytes are CREATE txs)

> The `logsbloom` field of the execution block header and the `logsbloom` field of the transaction receipt are now required to be an empty bytes array.

Reviewer 2: ```suggestion
The `logsBloom` field of the execution block header and the `logsBloom` field of the transaction receipt are now required to be an empty bytes array.
```

## Rationale

> This is a minimally disruptive way to remove the need to compute blooms from clients while not requiring any breaking consensus changes.

Reviewer 2: Well, this is a breaking consensus change, since it forces the bloom filter to be a constant which in almost all practical cases will not be the produced bloom filter when executing the block.

Author: Well, breaking on the level of "we're changing how this field is computed" but breaking the structure of the values that are RLP encoded into the serialized form of the header and receipt, right?

In practice, this mechanism is far too slow. Almost all dapps that access history end up doing so not through RPC calls to an Ethereum node (even a remote-hosted one), but through centralized extra-protocol services.  Not to mention, in a chain with relatively full blocks, the false positive rate in bloom filters has increased to the point of making bloom filters almost useless.

> This EIP proposes recognizing that reality, and removing bloom filters from the protocol since it places an unnecessary compute burden on execution clients computing essentially used bytes. Applications that need history querying would be encouraged to develop and use decentralized protocols that create provable log indexes using eg. ZK-SNARKs and incrementally-verifiable computation.

Author: Ugh, i can't type today.  Will revise my language.
